### DISCUSSION TITLE
NodeJs Tutorial Part 6 (RPC): RPC Client code does not work after the first request

### DISCUSSION THREAD

**Discussion by sfmohassel** [rabbitmq/rabbitmq-server#2758] [Other]
I've spent some time on
https://www.rabbitmq.com/tutorials/tutorial-six-javascript.html
and trying to wrap my head around the issue that the second request to some RPC Server won't work.

I've prepared a code for this tutorial, which is tested and I propose to use this code for this tutorial:

```javascript
const amqp = require('amqplib');
const uuid = require('uuid')
const EventEmitter = require('events').EventEmitter

class RabbitMqRpcClient {
    constructor() {
        /**
         * @type {amqp.Connection}
         */
        this.connection = null;

        /**
         * @type {amqp.Channel}
         */
        this.channel = null;

        /**
         * @type {EventEmitter}
         */
        this.eventEmitter = null

        /**
         * @type {string}
         */
        this.rpcQueue = ''

        /**
         * @type {string}
         */
        this.replyQueue = ''
    }

    /**
     * Connects and declares a queue for RPC
     * @param {string} rpcQueue
     * @param {amqp.Options.Connect} options
     * @returns {Promise<void>}
     */
    async connect(rpcQueue, options) {
        this.rpcQueue = rpcQueue
        const o = Object.assign({}, {
            protocol: 'amqp'
        }, options)
        this.connection = await amqp.connect(o)
        this.channel = await this.connection.createChannel()
        const replyQ = await this.channel.assertQueue('', {
            exclusive: true,
            durable: false,
            autoDelete: true
        })
        this.replyQueue = replyQ.queue


        this.eventEmitter = new EventEmitter()
        this.eventEmitter.setMaxListeners(0)

        this.channel.consume(this.replyQueue,
            msg => this.eventEmitter.emit(msg.properties.correlationId, msg.content),
            { noAck: true })
    }

    /**
     * Disconnects from RPC server
     * @returns {Promise<void>}
     */
    async disconnect() {
        this.eventEmitter.removeAllListeners()
        await this.channel?.close()
        await this.connection?.close()
        this.connection = null
        this.channel = null
    }

    /**
     * @param {Buffer} data
     * @returns {Promise<Buffer>}
     */
    call(data) {
        return new Promise((resolve, reject) => {
            try {
                // create correlation ID
                const correlationId = uuid.v4();
                this.eventEmitter.once(correlationId, buffer => {
                    try {
                        resolve(buffer)
                    } catch (error) {
                        reject(`Error while parsing the result: ${error}`)
                    }
                })
                this.channel.sendToQueue(this.rpcQueue, data, {
                    correlationId: correlationId,
                    replyTo: this.replyQueue
                })
            } catch (error) {
                reject(`Error while sending request: ${error}`)
            }
        })
    }
}

module.exports = RabbitMqRpcClient
```

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
Thank you for considering a contribution. I cannot reproduce:

``` shell
for i in {1..20}; do node src/rpc_client.js $i; done
 [x] Requesting fib(1)
 [.] Got 1
 [x] Requesting fib(2)
 [.] Got 1
 [x] Requesting fib(3)
 [.] Got 2
 [x] Requesting fib(4)
 [.] Got 3
 [x] Requesting fib(5)
 [.] Got 5
 [x] Requesting fib(6)
 [.] Got 8
 [x] Requesting fib(7)
 [.] Got 13
 [x] Requesting fib(8)
 [.] Got 21
 [x] Requesting fib(9)
 [.] Got 34
 [x] Requesting fib(10)
 [.] Got 55
 [x] Requesting fib(11)
 [.] Got 89
 [x] Requesting fib(12)
 [.] Got 144
 [x] Requesting fib(13)
 [.] Got 233
 [x] Requesting fib(14)
 [.] Got 377
 [x] Requesting fib(15)
 [.] Got 610
 [x] Requesting fib(16)
 [.] Got 987
 [x] Requesting fib(17)
 [.] Got 1597
 [x] Requesting fib(18)
 [.] Got 2584
 [x] Requesting fib(19)
 [.] Got 4181
 [x] Requesting fib(20)
 [.] Got 6765
```

**Comment by michaelklishin** [comment]
This example is substantially more involved and it's not obvious to me that it would be a better **tutorial**. A better demonstration of modern JavaScript, perhaps, but our tutorials focus on explaining RabbitMQ concepts with intentionally minimalistic code. Until this style of code becomes commonly requested, I think we will keep the tutorials as they are. We have years of evidence that they
achieve what they are meant to achieve: introduce beginners to the protocol and its features.

**Comment by sfmohassel** [comment]
Totally agree with you on 'keep it simple' for beginners.
I think you are not familiar with this article from 5 years ago:
[Link](https://www.google.com/amp/s/facundoolano.wordpress.com/2016/06/26/real-world-rpc-with-rabbitmq-and-node-js/amp/) 
The problem is that, functionally there's no problem with the current tutorial! But non-functionally the tutorial code is considered a bad practice:
- For each request, a new connection is made to RabbitMQ
- For each request, a new channel is created.
Why I'm proposing this?
Because .Net tutorial does not have above problems. In fact, `CancellationTokenSource` is just like the `EventEmitter` in my proposal.
If you consider my points applicable, I can reduce the modernism of the proposed code.
Anyway the reason I proposed, was that you do a great job

**Reply by sfmohassel** [reply]
Anyway I think my proposal would make the NodeJs code just like the .Net code, in terms of reusing connections and queues.

**Comment by lukebakken** [comment]
Thank you for the contribution. Questions:

- What is the minimum JavaScript (ECMAScript) version required to run this code?
- Are features like `const`, `async`, `EventEmitter` and `Promise` guaranteed to be available?
- What is the minimum version of nodejs required to run this code?

**Reply by sfmohassel** [reply]
This code needs at least ECMAScript 2017. All features are guaranteed to be available.
I've tested this code with `node v14.15.1`

**Comment by michaelklishin** [comment]
@saeidjoker OK, for this arguably more involved tutorial we can probably adopt this. Can you please submit a pull request to https://github.com/rabbitmq/rabbitmq-tutorials and another one to https://github.com/rabbitmq/rabbitmq-website (see under `site/tutorials`)? The website copy will have to be substantially updated to cover this version. Thank you.

**Reply by sfmohassel** [reply]
Yes of course. 
Unfortunately just today I've been diagnosed with COVID, So I'm basically really really tired. I'll do it next week hopefully after getting better. Man o man I wish nobody catches this virus...

These PRs would be an honor, because I owe a portion of my career to RabbitMQ (you guys) :-)

**Reply by michaelklishin** [reply]
Oh, I'm really sorry to hear that. Get well soon!
